cartridge/failover.lua

```diff
@@ -200,10 +206,146 @@ local function apply_config(mod)
+        vars.cache.is_leader = true
+        vars.cache.is_rw = true
+        return true
+    elseif vars.cache.is_vclockkeeper then
+        -- I'm already a vclockkeeper
+        return true
+    end
+
+    local deadline = fiber.time() + opts.timeout
+
+    -- Go to the external state provider
+    local session = assert(vars.client):get_session()
+    if session == nil or not session:is_alive() then
+        return nil, StateProviderError:new('State provider unavailable')
+    end
+
+    -- Query current vclockkeeper
+    -- WARNING: implicit yield
+    local vclockkeeper, err = session:get_vclockkeeper(replicaset_uuid)
+    fiber.testcancel()
+    if err ~= nil then
+        return nil, SwitchoverError:new(err)
+    end
+
+    if vclockkeeper == nil then
+        -- It's absent, no need to wait anyone
+        goto set_vclockkeeper
+    elseif vclockkeeper.instance_uuid == instance_uuid then
+        -- It's me already
+        goto set_vclockkeeper
+    end
+
+    do
+        -- Go to the vclockkeeper and query its LSN
+        local vclockkeeper_uuid = vclockkeeper.instance_uuid
+        local vclockkeeper_srv = topology_cfg.servers[vclockkeeper_uuid]
+        if vclockkeeper_srv == nil then
+            return nil, SwitchoverError:new('Alien vclockkeeper %q',
+                vclockkeeper.instance_uuid
+            )
+        end
+
+        local vclockkeeper_uri = vclockkeeper_srv.uri
+
+        local timeout = deadline - fiber.time()
+        -- WARNING: implicit yield
+        local vclockkeeper_info, err = errors.netbox_call(
+            pool.connect(vclockkeeper_uri, {wait_connected = false}),
+            '__cartridge_failover_get_lsn', {timeout},
+            {timeout = timeout + vars.options.NETBOX_CALL_TIMEOUT}
+        )
+        fiber.testcancel()
+
+        if vclockkeeper_info == nil then
+            return nil, SwitchoverError:new(err)
+        end
+
+        -- Wait async replication to arrive
+        -- WARNING: implicit yield
+        local timeout = deadline - fiber.time()
+        local ok = utils.wait_lsn(
+            vclockkeeper_info.id,
+            vclockkeeper_info.lsn,
+            vars.options.WAITLSN_PAUSE,
+            timeout
+        )
+        if not ok then
+            return nil, SwitchoverError:new(
+                "Can't catch up with the vclockkeeper"
+            )
+        end
+        fiber.testcancel()
+    end
+
+    -- The last one thing: persist our vclock
+    ::set_vclockkeeper::
+
+    -- WARNING: implicit yield
+    local ok, err = session:set_vclockkeeper(
+        replicaset_uuid,
+        instance_uuid,
+        box.info().vclock
+    )
+    fiber.testcancel()
+
+    if ok == nil then
+        return nil, SwitchoverError:new(err)
+    end
+
+    -- Hooray, instance is a legal vclockkeeper now.
+    vars.cache.is_vclockkeeper = true
+    vars.cache.is_leader = true
+    vars.cache.is_rw = true
+
+    log.info('Consistent switchover is successful')
```

Comment:
Consider adding "with LSN ..." (Not sure but I think it could be useful for diagnostic purposes)